In [1]:
# Import the necessary packages and data
from datetime import datetime
import numpy as np
import os
import pandas as pd

path = os.path.expanduser('~/Projects/capstone-two/data/raw/Harbor_Water_Quality.csv')
df = pd.read_csv(path, parse_dates=['Sample Date'])

C:\Users\Drew\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (5,6,10,11,12,13,20,21,26,27,35,38,40,51,53,59,60,61,64,65,67,68,76,82,83,88,94,95,97,98) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


There are dataype problems with certain columns, but before dealing with those, I first want to see what the dataframe looks like.

In [2]:
df.shape

(88366, 100)

In [3]:
df.head()

Sampling Location Duplicate Sample Sample Date Sample Time  \
0               BR5              NaN  2015-08-17     1:07 PM   
1                K2              NaN  1965-07-06     9:50 AM   
2                N8              NaN  1909-07-16     4:10 PM   
3                N9              NaN  1972-07-27     9:22 AM   
4               NC0              NaN  2014-02-26    10:45 AM   

  Weather Condition (Dry or Wet) Top Sample Temperature (ºC)  \
0                            Dry                       24.68   
1                            NaN                        22.8   
2                            NaN                          20   
3                            NaN                        17.2   
4                            Dry                        2.51   

  Bottom Sample Temperature (ºC)  Site Actual Depth (ft)  \
0                            NaN                    13.0   
1                           22.2                    21.0   
2                            NaN                     NaN   
3                           16.1                     NaN   
4                           2.46                    15.0   

   Top Sample Depth(ft)  Bottom Sample Depth (ft)  ...  \
0                   3.0                       NaN  ...   
1                   NaN                       NaN  ...   
2                   1.0                       NaN  ...   
3                   NaN                       NaN  ...   
4                   3.0                      12.0  ...   

  Oakwood BOD Top Sample (mg/L)  Oakwood BOD Bottom Sample(mg/L)  \
0                            NaN                             NaN   
1                            NaN                             NaN   
2                            NaN                             NaN   
3                            0.6                             0.6   
4                            NaN                             NaN   

  Oakwood  Total Suspended Solid Top Sample  (mg/L)  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  Oakwood Total Suspended Solid Bottom Sample (mg/L)  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

   Top Turbidity( Nephelometric Turbidity Units)  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   Bottom Turbidity YSI (Nephelometric Turbidity Units)  Sampling Comment  \
0                                                NaN                  NaN   
1                                                NaN                  NaN   
2                                                NaN                  NaN   
3                                                NaN                  NaN   
4                                                NaN           Light snow   

      Long        Lat       Type  
0 -73.8713  40.813667  Permanent  
1  -74.153  40.641167  Permanent  
2 -74.0455  40.606167  Permanent  
3 -73.9833  40.568333  Permanent  
4 -73.9316  40.714961  Permanent  

[5 rows x 100 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88366 entries, 0 to 88365
Data columns (total 100 columns):
Sampling Location                                                                   88365 non-null object
Duplicate Sample                                                                    2929 non-null object
Sample Date                                                                         88365 non-null datetime64[ns]
Sample Time                                                                         84504 non-null object
Weather Condition (Dry or Wet)                                                      32498 non-null object
Top Sample Temperature (ºC)                                                         69920 non-null object
Bottom Sample Temperature (ºC)                                                      69147 non-null object
Site Actual Depth (ft)                                                              60162 non-null float64
Top Sample Depth(ft)              

In [5]:
df.describe()

Site Actual Depth (ft)  Top Sample Depth(ft)  Bottom Sample Depth (ft)  \
count            60162.000000          27969.000000              27475.000000   
mean                34.193417              3.089239                 31.046695   
std                 21.903431              1.076717                 19.888996   
min                  0.000000              0.000000                  1.000000   
25%                 18.000000              3.000000                 16.400000   
50%                 28.000000              3.000000                 24.000000   
75%                 46.000000              3.280000                 41.000000   
max                230.000000             23.000000                150.000000   

       CTD (conductivity, temperature, depth profiler) Top Dissolved Oxygen (mg/L)  \
count                                       20367.000000                             
mean                                            6.950473                             
std                                             2.898322                             
min                                            -1.120000                             
25%                                             5.110000                             
50%                                             6.470000                             
75%                                             8.210000                             
max                                            28.560000                             

       CTD (conductivity, temperature, depth profiler) Bottom Dissolved Oxygen (mg/L)  \
count                                       18812.000000                                
mean                                            6.069471                                
std                                             2.856763                                
min                                            -9.990000                                
25%                                             4.390000                                
50%                                             5.660000                                
75%                                             7.240000                                
max                                            29.670000                                

       Percentage O2 Saturation Top Sample  \
count                         22129.000000   
mean                             81.811070   
std                              30.539045   
min                             -15.900000   
25%                              64.560000   
50%                              79.110000   
75%                              94.850000   
max                             374.220000   

       Percentage O2 Saturation Bottom sample  \
count                            20618.000000   
mean                                69.511316   
std                                 26.546844   
min                                -51.670000   
25%                                 55.572500   
50%                                 70.280000   
75%                                 83.700000   
max                                283.240000   

       Light Trans (% transparency) Top Sample  \
count                             16788.000000   
mean                                 58.964913   
std                                  20.532792   
min                                 -22.860000   
25%                                  49.980000   
50%                                  64.065000   
75%                                  72.920000   
max                                 131.730000   

       Light Trans (% transparency) Bottom Sample  Top Fluorometer (mg/m3)  \
count                                15781.000000             18555.000000   
mean                                    58.450028                15.080308   
std                                     22.295995                23.359770   
min                                    -41.970000               -99.000000   
25

In [6]:
len(df['Sampling Location'].unique())

5132

In [7]:
df['Sampling Location'].unique()

array(['BR5', 'K2', 'N8', ..., 'NR4-M1', 'NR1-M3', 'NR2-B2'], dtype=object)

There are 230 distinct locations, but I have a list of New York City Harbor Water Quality Survey Stations that is much shorter. The sites not listed in the following list are from older sampling events.

In [8]:
survey_stations = ['K1', 'K2', 'K3', 'K4', 'K5', 'K5A', 'K6',
                  'N1', 'N3B', 'N4', 'N5', 'N6', 'G2', 'N7', 'N8',
                  'N9', 'N16', 'NR1', 'E2', 'E4', 'E6', 'E7', 'E8',
                  'E10', 'E11', 'E12', 'E13', 'E14', 'E15', 'J1', 
                  'J2', 'J3', 'J5', 'J7', 'J8', 'J9A', 'J10', 'J11',
                  'J12', 'JA1', 'N9A', 'H3', 'J14', 'J16', 'AC1',
                  'AC1', 'AC2', 'BB2', 'BB4', 'BR1', 'BR3', 'BR5',
                  'CIC2', 'CIC3', 'F1', 'F5', 'FB1', 'FLC1', 'FLC2',
                  'GB1', 'GC3', 'GC4', 'GC5', 'GC6', 'HC1', 'HC2', 
                  'HC3', 'HR1', 'HR2', 'HR03', 'LN1', 'NC0', 'NC1',
                  'NC2', 'NC3', 'PB2', 'PB3', 'SP1', 'SP2', 'WC1',
                  'WC2', 'WC3'
                  ]
df = df[df['Sampling Location'].isin(survey_stations)]
df = df.reset_index(drop=True)

I don't want to look at data that is too old, so I am going to limit my dataset to readings from this millenium.

In [9]:
df = df[df['Sample Date'] >= datetime(2000, 1, 1)]
df = df.reset_index(drop=True)

I want to drop certain columns that do not have enough values or are not important to my question.

In [10]:
for col in df.columns:
    if 'Oakwood' in col:
        df = df.drop(columns=col)

#    elif 'Bottom' in col:
#        df = df.drop(columns=col)
    elif len(df[df[col].notnull()]) < 10000:
        df = df.drop(columns=col)
df = df.reset_index(drop=True)

Next I want to fix problematic columns

In [11]:
# Drop columns that I don't need
df = df.drop(['Current Direction (Current Direction)', 'Wind Direction (Wind Direction)',
              'Current Speed (knot)', 'Wind Speed (mph)', 'Sea State ', 'Type',
              'Enterococcus Top Sample Less Than or Greater Than Result'], axis=1)

In [12]:
# Fix 'Weather Condition' column
df['Weather Condition (Dry or Wet)'].unique()
df = df.replace(['Dry', 'Wet'], ['D', 'W'])

In [13]:
# Check which columns are numeric and create a list of object columns
obj_cols = []
ok_obj_cols = ['Sampling Location', 'Sample Date', 'Sample Time', 'Weather Condition (Dry or Wet)']
for col in df:
    if col not in ok_obj_cols:
        try:
            df[col] = pd.to_numeric(df[col])
        except:
            print(col)
            obj_cols.append(col)

Top Sample Photosysthetically Active Radiation Reference (400-700nm light (uE/S)
Top Sample Par (uE/S m2)
Secchi Depth (ft)
Bottom PH
Top Fecal Coliform Bacteria (Cells/100mL)
Top Enterococci Bacteria (Cells/100mL)
Top Nitrate/Nitrite (mg/L)
Top Ammonium (mg/L)
Top Ortho-Phosphorus (mg/L)
Top Silica (mg/L)
Total Phosphorus(mg/L)
Bottom Total Suspended Solid (mg/L)
Top Active Chlorophyll 'A' (µg/L)
Long
Lat


In [14]:
# Fix Bottom TSS
df['Bottom Total Suspended Solid (mg/L)'] = df['Bottom Total Suspended Solid (mg/L)'].replace('.', np.nan)

In [15]:
# Fix Secchi Disk
df['Secchi Depth (ft)'] = df['Secchi Depth (ft)'].replace('3..5', '3.5')

In [16]:
for i in df['Long']:
    try:
        pd.to_numeric(i)
    except:
        df['Long'] = df['Long'].replace(i, np.nan)

In [17]:
# Fix Bottom PH
df['Bottom PH'] = df['Bottom PH'].replace('N', np.nan)

In [18]:
# Fix Top Active Chlorophyl
df['Top Active Chlorophyll \'A\' (µg/L)'] = df['Top Active Chlorophyll \'A\' (µg/L)'].replace('.', np.nan)

In [19]:
# Preprocess Top Fecal Col
df['Top Fecal Coliform Bacteria (Cells/100mL)'] = df['Top Fecal Coliform Bacteria (Cells/100mL)'].replace('TNTC', np.nan)

In [20]:
# Fix bacteria
bacteria = ['Top Fecal Coliform Bacteria (Cells/100mL)',
           'Top Enterococci Bacteria (Cells/100mL)']
def fix_commas(string):
    '''Removes commas from strings representing numbers
    with values in the thousands'''
    try:
        string = pd.to_numeric(string)
    except:
        string = string.replace(',', '')
    return string

for col in bacteria:
    try:
        df[col] = df[col].apply(fix_commas)
        df[col] = pd.to_numeric(df[col])
    except:
        print(col)

In [21]:
# fix some of total phos
df['Total Phosphorus(mg/L)'] = df['Total Phosphorus(mg/L)'].replace(['0..454', '0,968'], ['0.454', '0.968'])

In [22]:
df['Top Ortho-Phosphorus (mg/L)'] = df['Top Ortho-Phosphorus (mg/L)'].replace('.', np.nan)

In [23]:
# fix < sign
less_than_cols = ['Top Nitrate/Nitrite (mg/L)', 'Top Ammonium (mg/L)',
                 'Top Ortho-Phosphorus (mg/L)', 'Top Silica (mg/L)',
                  'Total Phosphorus(mg/L)' 
                 ]
def drop_less(string):
    '''Removes commas from strings representing numbers
    with values in the thousands'''
    try:
        string = pd.to_numeric(string)
    except:
        string = pd.to_numeric(string.replace('<', ''))
    return string


for col in less_than_cols:
    try:
        df[col] = df[col].apply(drop_less)
        df[col] = pd.to_numeric(df[col])
    except:
        print(col)

In [24]:
def col_strip(string):
    '''Strips strings of trailing spaces and returns
    stripped string or numeric value'''
    if type(string) == str:
        string = string.replace(' ', '')
    return string

In [25]:
#df.info()

In [26]:
for i, lat in enumerate(df['Lat']):
    try:
        pd.to_numeric(lat)
    except:
        lat_long = lat.split(',')
        if len(lat_long) == 2:
            df.loc[:, 'Lat'][i] = lat_long[0]
            if len(lat_long[1]) > 1:
                df.loc[:, 'Long'][i] = lat_long[1]


C:\Users\Drew\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys
C:\Users\Drew\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [27]:
for col in df:
    if col != 'Sample Date':
        try:
            df[col] = df[col].apply(col_strip)
            df[col] = pd.to_numeric(df[col])
        except:
            print(col)

Sampling Location
Sample Time
Weather Condition (Dry or Wet)


In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31005 entries, 0 to 31004
Data columns (total 47 columns):
Sampling Location                                                                   31005 non-null object
Sample Date                                                                         31005 non-null datetime64[ns]
Sample Time                                                                         29217 non-null object
Weather Condition (Dry or Wet)                                                      30601 non-null object
Top Sample Temperature (ºC)                                                         18996 non-null float64
Bottom Sample Temperature (ºC)                                                      17733 non-null float64
Site Actual Depth (ft)                                                              27886 non-null float64
Top Sample Depth(ft)                                                                18939 non-null float64
Bottom Sample Depth (ft)       

In [29]:
df.describe()

Top Sample Temperature (ºC)  Bottom Sample Temperature (ºC)  \
count                 18996.000000                    17733.000000   
mean                     18.703483                       18.002141   
std                       6.801709                        6.523780   
min                      -1.740000                       -1.870000   
25%                      15.737500                       15.000000   
50%                      21.190000                       20.340000   
75%                      23.570000                       22.830000   
max                      30.010000                       29.230000   

       Site Actual Depth (ft)  Top Sample Depth(ft)  Bottom Sample Depth (ft)  \
count            27886.000000          18939.000000              17836.000000   
mean                28.630030              3.380010                 32.244005   
std                 21.298745              0.749955                 20.358388   
min                  0.020000              0.700000                  1.000000   
25%                 14.000000              3.000000                 17.000000   
50%                 21.000000              3.000000                 25.000000   
75%                 40.000000              4.000000                 43.000000   
max                230.000000             12.000000                107.000000   

       Top Salinity  (psu)  Bottom Salinity  (psu)  Top Conductivity (S/m)  \
count         18968.000000            17720.000000            18758.000000   
mean             22.866737               24.809579                3.837574   
std               5.269741                3.794550                4.396670   
min               0.120000                0.200000                0.020000   
25%              21.160000               23.060000                2.730000   
50%              24.060000               25.240000                3.370000   
75%              26.220000               27.270000                3.820000   
max              34.760000               44.730000               43.960000   

       Bottom Conductivity (S/m)  \
count               17528.000000   
mean                    4.090200   
std                     4.665498   
min                     0.000000   
25%                     2.980000   
50%                     3.500000   
75%                     3.900000   
max                    62.560000   

       CTD (conductivity, temperature, depth profiler) Top Dissolved Oxygen (mg/L)  \
count                                       16217.000000                             
mean                                            6.851332                             
std                                             2.871964                             
min                                            -1.120000                             
25%                                             5.080000                             
50%                                             6.410000                             
75%                                             8.080000                             
max                                            28.560000                             

       ...  Top Ortho-Phosphorus (mg/L)  Top Total Kjeldhal Nitrogen (mg/L)  \
count  ...                 19297.000000                        28533.000000   
mean   ...                     0.165340                            1.183278   
std    ...                     0.275925                            1.865418   
min    ...                     0.000000                            0.033000   
25%    ...                     0.070000                            0.608000   
50%    ...                     0.115000                            0.858000   
75%    ...                     0.164000                            1.270000   
max    ...                     5.210000                           44.800000   

       Top Silica (mg/L)  Total Phosphorus(mg/L)  \
count       28507.000000            28554.000000   
mean            2.151428   

There are some longitude and latitudes that are switched.

In [30]:
for i, long in enumerate(df['Long']):
    if long > 0:
        lat = long
        long = df.iloc[i]['Lat']
        df.iloc[i]['Lat'] = lat
        df.iloc[i]['Long'] = long

C:\Users\Drew\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
C:\Users\Drew\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [31]:
outpath = os.path.expanduser('~/Projects/capstone-two/data/processed/Clean_Harbor_Water_Quality.csv')
df.to_csv(outpath)